### DISCUSSION TITLE
personalized workouts Treadmill managed automatically by HR rate indicating the BPM instead the PID HR

### DISCUSSION THREAD

**Discussion by juanrancho** [cagnulein/qdomyos-zwift#2276] [General]
Hi,

I know that you can do workouts where the threadmill speed is adapted automatically depending the Heart Rate zone that you have indicated. 
If I am not wrong, the process is:

1.- introduce your Heart Rate data:
   1.1- Heart Rate Options/Heart Rate zone options: You can activate "heart rate max override" to intro your personal max HR, and then you can indicate the 4 HR zones in % of your max HR.

2.- To especify that you want to do a workout based in HR zone: Training programs options/PID on Heart zone and indicate a zone instead "disabled".

3.- start a workout

If this is correct, it´s ok for me.

My questions

1.- is if it is possible to build a preconfigured workout (.xml or .zwo) based in HR zones, and in positive case, if is it possible to indicate the BPM instead the HR Zone.

I have view in (https://github.com/cagnulein/qdomyos-zwift/blob/5bfa0bba16cfba11108182afd085d2b0fd066293/src/trainprogram.cpp#L1349) that there are several parameters like:
- zonehr
- hrmin
- hrmax
- looptimehr
can you explain how they work? 

2.- is it possible to build a preconfigured workout (xml or .zwo) based in power? (I am using Stryd from 1 week)
How can I do it? based in power zones? based in % of the CP (Critical Power)? can I indicate exactly the power?

thanks.

**Comment by cagnulein** [comment]
> I know that you can do workouts where the threadmill speed is adapted automatically depending the Heart Rate zone that you have indicated. If I am not wrong, the process is:
> 
> 1.- introduce your Heart Rate data: 1.1- Heart Rate Options/Heart Rate zone options: You can activate "heart rate max override" to intro your personal max HR, and then you can indicate the 4 HR zones in % of your max HR.

yes that's right.

> 2.- To especify that you want to do a workout based in HR zone: Training programs options/PID on Heart zone and indicate a zone instead "disabled".

yes and also you can change it with the PID hr tile on the fly!


> 3.- start a workout
> 
> If this is correct, it´s ok for me.

yes correct!


> My questions
> 
> 1.- is if it is possible to build a preconfigured workout (.xml or .zwo) based in HR zones, and in positive case, if is it possible to indicate the BPM instead the HR Zone.

yes it's possible!


> I have view in (
> 
> https://github.com/cagnulein/qdomyos-zwift/blob/5bfa0bba16cfba11108182afd085d2b0fd066293/src/trainprogram.cpp#L1349
> 
> ) that there are several parameters like:
> * zonehr
> * hrmin
> * hrmax
> * looptimehr
>   can you explain how they work?


zonehr is the zone that you want to mantain. 
hrmin and hrmax defines intead a range of hr instead of zonehr. you can't use zonehr with hrmin and hrmax. or you use zonehr, or hrmin and hrmax
looptimehr is the time in second that QZ change the speed. For example if it set to 5, qz will check your hr every 5 seconds and it will try to change the speed accordingly every 5 second. 


> 2.- is it possible to build a preconfigured workout (xml or .zwo) based in power? (I am using Stryd from 1 week) How can I do it? based in power zones? based in % of the CP (Critical Power)? can I indicate exactly the power?
> 
> thanks.

yes! it's what I did the last day here https://github.com/cagnulein/qdomyos-zwift/commit/6cd52efe77c15b440c3c1a1e78b123dc3ca19b9c

i have a stryd too!
for now i did it using the wattage directly and it's basing on the wattage calculation from QZ but in case i can create a stryd lookup table too if you are interested. And also I can add the powerzone tag as well in the xml.

I suggest to create a ticket about this instead of a discussion for this particula task :)

**Reply by cagnulein** [reply]
@juanrancho i just added the erg table for the stryd too if you want to play with it! it's on the 2.16.52

**Comment by juanrancho** [comment]
thank you @cagnulein , how can I upgrade to 2.16.52 (I am beta tester but I am in 2.16.47).
And, sorry for the naive question, I am not sure to understand how is the way to use stryd for my power based workouts.
My aim is design workouts based in power like: (I am a very popular runner, nothing professional)(but very professional in training demand because I am a IT lover)(example: my best 10K = 4:22mins/Km)
- 5 mins (warm up power)(equivalent to 7 mins/km)
- 10 mins (increasing power from warm up power to 10 km/h)
- intervals 10x(1 min 11 km/h - 1 min 15 km/h)
- 10 mins (power equivalent to 5 mins/km)(12km/h)
- 5 mins 8 km/h 
and all designed by power from stryd.
how would the xml?

thanks in advance.

PS: I really love your app. I'm a telecommunications engineer (working in a multinational IT company) and I'm available to give you feedback on all the things that a very demanding non-professional runner/cyclist (I think there are a huge amount of people waiting for things like what I would like, even if they don't know it) like me would like to have.

**Comment by cagnulein** [comment]
> thank you @cagnulein , how can I upgrade to 2.16.52 (I am beta tester but I am in 2.16.47). 

i didn't push yet to the play store because I would like to test it first. The apk anyway it's available here https://github.com/cagnulein/qdomyos-zwift/actions/runs/8577205250/artifacts/1390065620

What I did in this version is: qz is learning the power curve from the stryd. In order to do it, do a 10 minute session changing the most often speed used for you, in this way qz will learn the power and speed and the inclination from both stryd and treadmill and when the user will ask for a target power, qz will set it! the power tag in the xml is already available in this build and it's expressed in wattage (so 200 for 200W)

Google is very slow in these days, reviewing the app on the store, even the beta ones.  In order to test it you need first to remove the current QZ version on your phone, install the one that you downloaded.
This version is a 15 minutes trial (you can restart it as many times as you want) and I can unlock it if you want, send me an email to roberto.viola83@gmail.com mentioning this ticket and that you are on android.
Let me know if you have any questions
I suggest to use a spare android for this test 

> And, sorry for the naive question, I am not sure to understand how is the way to use stryd for my power based workouts. My aim is design workouts based in power like: (I am a very popular runner, nothing professional)(but very professional in training demand because I am a IT lover)(example: my best 10K = 4:22mins/Km)
> 
> * 5 mins (warm up power)(equivalent to 7 mins/km)
> * 10 mins (increasing power from warm up power to 10 km/h)
> * intervals 10x(1 min 11 km/h - 1 min 15 km/h)
> * 10 mins (power equivalent to 5 mins/km)(12km/h)
> * 5 mins 8 km/h
>   and all designed by power from stryd.
>   how would the xml?

I didn't do this yet, I will do next week so we will have a complete stryd integration

> 
> thanks in advance.
> 
> PS: I really love your app. I'm a telecommunications engineer (working in a multinational IT company) and I'm available to give you feedback on all the things that a very demanding non-professional runner/cyclist (I think there are a huge amount of people waiting for things like what I would like, even if they don't know it) like me would like to have.

Great!

**Comment by juanrancho** [comment]
I have one doubt:
 - is possible design a workout (.xml) mixing sements by speed and segments by HR?
 - Let me put an example:

```
<?xml version="1.0" encoding="UTF-8"?>
<rows>
	<row forcespeed="1" speed=8" inclination="0" fanspeed="0" duration="00:00:30" />
	<row forcespeed="1" speed="9" inclination="0" fanspeed="0" duration="00:00:30" />
	<row duration="00:01:00" hrmin="120" hrmax="130" fanspeed="0" maxspeed="12" looptimehr="3"/>
	<row duration="00:01:00" hrmin="90" hrmax="100" fanspeed="0" maxspeed="9" looptimehr="3"/>
	<row forcespeed="1" speed="4" inclination="0" fanspeed="0" duration="00:01:00" />
	
</rows>
```

As you can see, I use 2 rows by speed and then 2 more by HR to end with a row by speed.

would it work?

thanks.

**Reply by cagnulein** [reply]
Yes! In order to be sure, in the speed lines, put also hrmin and hrmax to 0

let me know

**Comment by juanrancho** [comment]
thank you @cagnulein,
other doubt:
- is possible to design a workout (.xml) for bike by HR Zones?
- the resistance (and power) of the bike should be controlled automatically to maintain the hrmin and hrmax.
thanks.

**Reply by cagnulein** [reply]
Yes @juanrancho same as treadmill!

**Comment by cagnulein** [comment]
> thank you @cagnulein , how can I upgrade to 2.16.52 (I am beta tester but I am in 2.16.47). And, sorry for the naive question, I am not sure to understand how is the way to use stryd for my power based workouts. My aim is design workouts based in power like: (I am a very popular runner, nothing professional)(but very professional in training demand because I am a IT lover)(example: my best 10K = 4:22mins/Km)
> 
> * 5 mins (warm up power)(equivalent to 7 mins/km)
> * 10 mins (increasing power from warm up power to 10 km/h)
> * intervals 10x(1 min 11 km/h - 1 min 15 km/h)
> * 10 mins (power equivalent to 5 mins/km)(12km/h)
> * 5 mins 8 km/h
>   and all designed by power from stryd.
>   how would the xml?
> 
> thanks in advance.
> 
> PS: I really love your app. I'm a telecommunications engineer (working in a multinational IT company) and I'm available to give you feedback on all the things that a very demanding non-professional runner/cyclist (I think there are a huge amount of people waiting for things like what I would like, even if they don't know it) like me would like to have.

@juanrancho  I did it in the 2.16.53 !

these are the files to test the new features

[qzramptest.zip](https://github.com/cagnulein/qdomyos-zwift/files/14904047/qzramptest.zip)

in the qzramptest you can see the new xml tag for do a speed ramp
in the qzpowerzonetest you can see how i translated your workout into xml file

Let me know!

**Reply by cagnulein** [reply]
ok course @juanrancho if you want I can handle also repetition in one line only!

**Comment by cagnulein** [comment]
@juanrancho any news?

**Comment by juanrancho** [comment]
I had no time to test your Last xmls.
I expect to do it in the weekend
________________________________

**Comment by juanrancho** [comment]
Hi @cagnulein , I am going to test the .xmls but in the meantime, I have one doubt related the "powerzone" number.

How is it calculated?

I mean, as you know, my aim is to use

- stryd for power
- my Polar heart rate monitor
- my nordictrack T10.0 threadmill

, then, I have 3 comments:

1.- do I need to change any parameter in the QZ app for that? (I supose that I need to tell the app from where the power measure come)

2.- once the point 1 solved, is not possible configure my .xml file using a power number (ex 230 W) instead a "powerzone" ?. In the negative case, how is calculated the "powerzone" number from the power (W) number? (I supose that QZ use a kind of power from it´s internal calculations but I don´t know how is it translated to stryd measures).

3.- Related with the graphics, my aim would be able to view 3 types of measures when I am training:
- 3.1- the real measures of my performance (ex: threadmill speed and inclination, power from stryd, heart rate,...)
- 3.2 - the measures that I programed in the current interval of training and the time remaining to finish that interval
- 3.3 - the measures of the next interval of training including the duration. 
It would be better to view a graphic with the entire programmed training but I don´t know if you are working on that.
As an example, I attach a screenshot from "trainerday" app where you can see the figures in number format and a graphic representation where you can see the entire programmed training and over it, my real performance of heart rate and power (in this case Trainerday is for bike). (I will attach the screenshot in a couple of minutes when I was training)
thanks in advance

**Reply by cagnulein** [reply]
> Hi @cagnulein , I am going to test the .xmls but in the meantime, I have one doubt related the "powerzone" number.

OK remember to use the apk that I linked above

> 
> 
> 
> How is it calculated?

1.00 equals to 100% of your CP of the qz setting 

> 
> 
> 
> I mean, as you know, my aim is to use
> 
> 
> 
> - stryd for power
> 
> - my Polar heart rate monitor
> 
> - my nordictrack T10.0 threadmill
> 
> 
> 
> , then, I have 3 comments:
> 
> 
> 
> 1.- do I need to change any parameter in the QZ app for that? (I supose that I need to tell the app from where the power measure come)


Yes power sensor under accessories to stryd and also the CP on the general settings 

> 
> 
> 
> 2.- once the point 1 solved, is not possible configure my .xml file using a power number (ex 230 W) instead a "powerzone" ?. In the negative case, how is calculated the "powerzone" number from the power (W) number? (I supose that QZ use a kind of power from it´s internal calculations but I don´t know how is it translated to stryd measures).

Yes you can use the tag power to set directly the wattage (without the W in the number)

> 
> 
> 
> 3.- Related with the graphics, my aim would be able to view 3 types of measures when I am training:
> 
> - 3.1- the real measures of my performance (ex: threadmill speed and inclination, power from stryd, heart rate,...)
> 
> - 3.2 - the measures that I programed in the current interval of training and the time remaining to finish that interval
> 
> - 3.3 - the measures of the next interval of training including the duration. 
> 
> It would be better to view a graphic with the entire programmed training but I don´t know if you are working on that.
> 
> As an example, I attach a screenshot from "trainerday" app where you can see the figures in number format and a graphic representation where you can see the entire programmed training and over it, my real performance of heart rate and power (in this case Trainerday is for bike). (I will attach the screenshot in a couple of minutes when I was training)
> 
> thanks in advance


I can add real-time charts yes, let's see first if the training program is working.

Remember to run to some different paces to calibrate qz first, so it can know your power curve

**Reply by cagnulein** [reply]
BTW I can't see the charts that you mentioned

**Reply by juanrancho** [reply]
![Screenshot_20240428_101117_Spotify.jpg](https://github.com/cagnulein/qdomyos-zwift/assets/163849058/b9936071-18f3-4b05-9fd2-fe9f57b6da00)

**Reply by juanrancho** [reply]
Graphic Red is health rate, White is real poder,  all over programmed trainninggraph

**Reply by cagnulein** [reply]
Yeah sure it's doable, you can already have an idea clicking on the chart icon when the training program is loaded

**Reply by juanrancho** [reply]
Sorry, but, I dont´know what do you refer with "chart icon", I can not view any chart icon (of perhaps I am not able to identify it).
Can you share any screenshots?
thanks.

**Reply by cagnulein** [reply]
@juanrancho it will appear on the left top corner when you load a power zone workouts. I will check if I enabled it for treadmill workouts too. Maybe I didn't

**Comment by juanrancho** [comment]
one doubt: I am beta tester and I think I am in 2.16.52 version.
You mentioned 2.16.53 version, is it available? (Android)

**Reply by juanrancho** [reply]
other one:
is possible to use the parameters "powerfrom" and "powerto" in the same way than "powerzonefrom" and "powerzoneto"?.
this would be to use power directly instead a % of power

**Reply by cagnulein** [reply]
@juanrancho it's in the closed beta, I can invite you. Send me an email to Roberto.viola83 at Gmail.com indicating this ticket

**Comment by cagnulein** [comment]
> other one: is possible to use the parameters "powerfrom" and "powerto" in the same way than "powerzonefrom" and "powerzoneto"?. this would be to use power directly instead a % of power

Yes
